blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md

```diff
@@ -0,0 +1,397 @@
+
+## Genomes / author: jamie
+Genomes in MABE2 are just repositories of mutable information.  The most basic kinds are vectors of some type, and (unlike in MABE), they aren't hidden inside Organisms which manage everything about them, but they are actively added to them and manage themselves.  This means that to mutate a Genome, any Genome, the Organism just tells the Genome to mutate itself and the Genome does.  No longer will the Organism run all the mutation functions itself, now it just delegates.  This, critically, means that Organisms now no longer need to know what kinds of Genomes they have at all.
+
+Genomes in MABE2 are also accessed differently -- regardless of what type the Genome is, it can be read from and written to as if it were made of doubles, ints, bytes, or bits.  They can also be read from and written to in bulk, and they can be "scaled", where what you see is automatically scaled to whatever the Genome's actual range of values is.  
+
+Moreover, Genomes are often handled through the Genome::Head class, which allows for easy and abstract use of Genomes.  Heads move along a Genome, and through a Head, any Genome operation can be accomplished, including searching (for values, ranges, multiple values, and ranges of multiple values) and all aforementioned reads and writes.  
+
+This means that it is often beneficial to pass Genomes as Heads if you want to be as generic as possible while still allowing full functionality of the Genome.
+  
+The most important part of a Genome is the mutate function.  Here's an example of a very simple mutate function for a vector of bits:
+  
+```cpp
+  
+  size_t Mutate(emp::Random & random) override { // receives a random number generator
+      // if we haven't configured the mutation sites yet, do so and set up the binomial distribution
+      if (mut_sites.GetSize() != data.size()) { 
+        mut_dist.Setup(mut_p, data.size());
+        mut_sites.resize(data.size());
+      }
+      // get the number of mutations
+      const size_t num_muts = mut_dist.PickRandom(random); 
+      
+
+      if (num_muts == 0) return 0;
+      if (num_muts == 1) {
+        const size_t pos = random.GetUInt(data.size());
+        data.Toggle(pos); // swaps the bit
+        return 1;
+      }
+
+      // Only remaining option is num_muts > 1.
+      mut_sites.Clear();
+      for (size_t i = 0; i < num_muts; i++) {
+        const size_t pos = random.GetUInt(data.size());
+        if (mut_sites[pos]) { --i; continue; }  // duplicate position; try again.
+        mut_sites.Set(pos); // sets each site of mutation to 1
+      }
+      data ^= mut_sites; // computes an exclusive or with the vector of bits and the mutation position vector
+
+      return num_muts;
+    }
+  
+```  
+
+Naturally, there are many more ways to mutate a Genome, particuarly if the Genome is a vector of integers or doubles.  If you want to make a Genome that mutates in a particular fashion, this is the way you'll have to go about it. 
```

Comment:
Jamie can answer this better than me, but I reworded it to make it less confusing.

Reply:
I elaborated on that further in my earliest draft, but I ended up scrapping that section since it got too long and too off-topic, not to mention technical beyond the scope of the blog post.  Mitchel did a great job in clarifying it.